### PR TITLE
HTTP Headers rework.

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

**Comment by dkorolev** [comment]
@sompylasar : FYI. Hope you like the design. :-)
@mzhurovich : Bad news 1/2: I had to touch EventCollector and Midichlorians. Bad news 2/2: Will need to fix Mac's native HTTP.
@dkorolev: Accepting headers without a space after them in HTTP is not done yet. The rest is tested end to end.

Thanks,
Dima

**Comment by mzhurovich** [comment]
12 марта 2016 г., в 1:17, Dima <notifications@github.com> написал(а):

@sompylasar <https://github.com/sompylasar> : FYI. Hope you like the
design. :-)
@mzhurovich <https://github.com/mzhurovich> : Bad news 1/2: I had to touch
EventCollector and Midichlorians. Bad news 2/2: Will need to fix Mac's
native HTTP.
@dkorolev <https://github.com/dkorolev>: Accepting headers without a space
after them in HTTP is not done yet. The rest is tested end to end.

Thanks,
Dima

—
Reply to this email directly or view it on GitHub
<https://github.com/c5t/Current/pull/417#issuecomment-195574145>.

**Comment by mzhurovich** [comment]
@dkorolev No problem, I'll fix Mac code.

**Comment by dkorolev** [comment]
@mzhurovich You da man. I'd be more worried about `EventCollector` and `Midichlorians` though. Certainly not now, gotta grab a bite a do some prod work.

Thanks!!!

**Comment by mzhurovich** [comment]
@dkorolev :hamburger: :beer:

File: Bricks/net/http/headers/headers.h [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+          return false;
+        }
+      }
+      return lhs_length < rhs_length;
+    }
+  };
+
+  // Make sure `Cookies` are never accessed as regular headers.
+  static bool IsCookieHeader(const std::string& header) {
+    const char* kCookieHeaderName = "Cookie";
```
Comment (mzhurovich) [inline Bricks/net/http/headers/headers.h]: What about server-side `Set-Cookie`?

**Comment by mzhurovich** [comment]
LGTM++, this is really cool!

File: Bricks/net/http/headers/headers.h (hunk 1 above)
Comment (dkorolev) [inline Bricks/net/http/headers/headers.h]: It's set in two places where cookies are sent, not received. Let's sync up by voice before I merge it in, ready in 15.

**Pull request merged**